### ISSUE TITLE
Replication performance: PouchDB may be downloading more docs than it needs to

### ISSUE THREAD

**Issue by alxndrsn** [medic/medic-webapp#1780]
pouch is generating a query similar to this:

`/medic/_changes?timeout=25000&style=all_docs&heartbeat=10000&filter=medic%2Fdoc_by_place&_nonce=1450453351782`

the problem is in the `style` value:

> `style` _(string)_ – Specifies how many revisions are returned in the changes array. The default, `main_only`, will only return the current “winning” revision; `all_docs` will return all leaf revisions (including conflicts and deleted former conflicts).
> –_http://docs.couchdb.org/en/latest/api/database/changes.html_

Is there an obvious reason why pouch would need `all_docs` (as it does currently) instead of `main_only`?

**Comment by alxndrsn** [comment]
N.B. pouch supports `style=main_only` for [`.changes()`](http://pouchdb.com/api.html#changes), but not for [`.replicate()`](http://pouchdb.com/api.html#replication).

**Comment by estellecomment** [comment]
Couch, like pouch, doesn't have the "style=all_docs/main_only" option for replication, only for changes feed. 
http://docs.couchdb.org/en/1.6.1/json-structure.html#replication-settings

**Comment by estellecomment** [comment]
TLDR : main_only would be fine for the first sync : there's nothing on the client db, so it doesn't matter.
Once both client and server are editing docs, I think we'll need to know all_docs for handling conflicts. 

**The replication algo wants all the revisions** : 
From http://guide.couchdb.org/draft/conflicts.html, the algorithm for deciding which version wins in a conflict:
"Each revision includes a list of previous revisions. The revision with the longest revision history list becomes the winning revision. If they are the same, the _rev values are compared in ASCII sort order, and the highest wins."
So that might be why you need the whole list of revisions.

From a Couchbase doc (not couchdb, but the output of the replication should be the same) 
https://github.com/couchbase/couchbase-lite-ios/wiki/Replication-Algorithm#goal
It says replications includes "contents, attachments and histories" (histories = all_docs). 
Further down, step 4, calling the _changes feed : "Use the "?style=all_docs" URL parameter so that conflicting revisions will be included." (the latest revision of a doc could be a conflict : two revisions, with one that is declared winner by the replication algo. But the two are kept, because your app might want to keep the loser instead, or make a merge of both.)

**Conclusion : we can't change all_doc -> main_only without fiddling with the replication algorithm**. 
That's a bad idea IMO : robust replication is what couch/pouch is all about, we're likely to get conflict problems.

For the initial replication though (when the client db is empty), we could do with main_only. 
**Couch people don't want to implement main_only replication because it's dangerous** : http://qnalist.com/questions/2288464/allow-style-main-only-on-the-source-feed-when-replicating

Other option is **use a bulk-get API, instead of replication, for the first sync**: GET the list of doc ids from the docs_by_place view, then get docs in bulk (or small chunks, or individually, whatever works).
http://docs.couchdb.org/en/1.6.1/api/database/bulk-api.html#
We could benchmark if that's faster than replication. Since the user would be waiting behind a splash screen, that would be an improvement.
(We should also make sure that the docs keep the same revision ids, so that the replication doesn't re-download them for nothing later, thinking they're different docs)

**Comment by estellecomment** [comment]
Closing, turns out Pouch is doing things right.
